bindings/go/input/dir/blob.go

```
@@ -0,0 +1,291 @@
+	// TODO:
+	// - Handle FollowSymlinks option.
+
+	// Pack directory contents as a tar archive.
+	reader, err := packDirToTar(ctx, dir.Path, &dir)
+	if err != nil {
+		return nil, fmt.Errorf("error producing blob for a dir input: %w", err)
+	}
+
+	// Wrap the tar archive in a ReadOnlyBlob.
+	mediaType := dir.MediaType
+	if mediaType == "" {
+		mediaType = DEFAULT_TAR_MIME_TYPE
+	}
+	var dirBlob blob.ReadOnlyBlob = inmemory.New(reader, inmemory.WithMediaType(mediaType))
+
+	// gzip the blob, if requested in the spec.
+	if dir.Compress {
+		dirBlob = compression.Compress(dirBlob)
+	}
+
+	return dirBlob, nil
+}
+
+// packDirToTar is the main function, which creates a tar archive from the contents of the specified directory.
+// It creates an instance of the virtual FileSystem based on the directory path, creates a tar writer and
+// triggers recursive packaging of the directory contents.
+func packDirToTar(ctx context.Context, path string, opt *v1.Dir) (_ io.Reader, err error) {
+	if path == "" {
+		return nil, fmt.Errorf("dir path must not be empty")
+	}
+
+	// Determine the base directory for relative paths in the tar archive.
+	baseDir := path
+	subDir := ""
+	if opt.PreserveDir {
+		// PreserveDir defines that the directory specified in the path field should be included in the blob.
+		baseDir = filepath.Dir(path)
+		subDir = filepath.Base(path)
+	}
+
+	// Create a new virtual FileSystem instance based on the provided directory path.
+	fileSystem, err := filesystem.NewFS(baseDir, os.O_RDONLY)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create filesystem while trying to access %v: %w", path, err)
+	}
+
+	// Create a pipe for streaming the tar data
+	pr, pw := io.Pipe()
+
+	// Start a goroutine to create the tar and write the data to the pipe.
+	go func() {
+		defer func() {
+			err = pw.Close()
+			if err != nil {
+				log.Base().Log(ctx, slog.LevelError, fmt.Sprintf("failed to close PipeWriter: %s", err.Error()))
```

why do you log here if you already return the close error?

---

This is how we discussed it last time. It is a `defere` in a `goroutine`. The error is not returned. I now removed the named return value to avoid confusion.

---

Done! Thanks!